rclpy/src/rclpy/_rclpy.c

```diff
@@ -2521,145 +2449,6 @@ rclpy_time_since_last_call(PyObject * module, PyObject * args)
-    return;
-  }
-
-  rcl_ret_t ret = rcl_subscription_fini(&(sub->subscription), sub->node);
-  if (RCL_RET_OK != ret) {
-    // Warning should use line number of the current stack frame
-    int stack_level = 1;
-    PyErr_WarnFormat(
-      PyExc_RuntimeWarning, stack_level, "Failed to fini subscription: %s",
-      rcl_get_error_string().str);
-  }
-  PyMem_Free(sub);
-}
-
-/// Create a subscription
-/**
- * This function will create a subscription for the given topic name.
- * This subscription will use the typesupport defined in the message module
- * provided as pymsg_type to send messages over the wire.
- *
- * On a successful call a list with two elements is returned:
- *
- * - a Capsule pointing to the pointer of the created rcl_subscription_t * structure
- * - an integer representing the memory address of the created rcl_subscription_t
- *
- * Raises ValueError if the capsules are not the correct types
- * Raises RuntimeError if the subscription could not be created
- *
- * \param[in] pynode Capsule pointing to the node to add the subscriber to
- * \param[in] pymsg_type Message module associated with the subscriber
- * \param[in] pytopic Python object containing the topic name
- * \param[in] pyqos_profile QoSProfile Python object for this subscription
- * \return list with the capsule and memory address, or
- * \return NULL on failure
- */
-static PyObject *
-rclpy_create_subscription(PyObject * module, PyObject * args)
-{
-  rclpy_module_state_t * module_state = (rclpy_module_state_t *)PyModule_GetState(module);
-  if (!module_state) {
-    // exception already raised
-    return NULL;
-  }
-  PyObject * pynode;
-  PyObject * pymsg_type;
-  PyObject * pytopic;
-  PyObject * pyqos_profile;
-
-  if (!PyArg_ParseTuple(args, "OOOO", &pynode, &pymsg_type, &pytopic, &pyqos_profile)) {
-    return NULL;
-  }
-
-  const char * topic = PyUnicode_AsUTF8(pytopic);
-  if (!topic) {
-    return NULL;
-  }
-
-  rclpy_handle_t * node_handle = PyCapsule_GetPointer(pynode, "rcl_node_t");
-  if (!node_handle) {
-    return NULL;
-  }
-  rcl_node_t * node = _rclpy_handle_get_pointer(node_handle);
-  if (!node) {
-    return NULL;
-  }
-
-  rosidl_message_type_support_t * ts = rclpy_common_get_type_support(pymsg_type);
-  if (!ts) {
-    return NULL;
-  }
-
-  rcl_subscription_options_t subscription_ops = rcl_subscription_get_default_options();
-
-  if (PyCapsule_IsValid(pyqos_profile, "rmw_qos_profile_t")) {
-    void * p = PyCapsule_GetPointer(pyqos_profile, "rmw_qos_profile_t");
-    rmw_qos_profile_t * qos_profile = p;
-    subscription_ops.qos = *qos_profile;
-    // TODO(jacobperron): It is not obvious why the capsule reference should be destroyed here.
-    // Instead, a safer pattern would be to destroy the QoS object with its own destructor.
-    PyMem_Free(p);
-    if (PyCapsule_SetPointer(pyqos_profile, Py_None)) {
-      // exception set by PyCapsule_SetPointer
-      return NULL;
-    }
```

Comment:
Am I correct that this isn't necessary after conversion to pybind11? I think the `py::capsule` takes care of it, right?

Reply:
Hmm I remember this - I forgot it wasn't fixed. The `rmw_qos_profile_t` conversion is a problem because it's allocating memory but not setting a pycapsule destructor. It's assuming that code here manually frees it, but that of course assumes a user passes it to one of these functions like `rclpy_create_subscription`.

https://github.com/ros2/rclpy/blob/ccc83acb5e534c00f5b04ee31ffde6028fb9ddf6/rclpy/src/rclpy/_rclpy.c#L4602

We should fix that first (make the rmw_qos_profile_t capsule have a destructor that is), and then we won't need this code block because the pycapsule will take care of it.

Reply:
This should be resolved by #700, and no changes should be necessary to this PR.